tests/gradients/core/test_hadamard_gradient.py

```diff
@@ -105,27 +105,307 @@ def cost12(x):
+        """Test that a ValueError is raised if more than one expectation value is used
+        with reversed and reversed-direct methods."""
+
+        tape = qml.tape.QuantumScript(
+            [qml.RX(0.543, 0), qml.RY(-0.654, 0)], [qml.expval(qml.Z(0)), qml.expval(qml.X(0))]
+        )
+        with pytest.raises(NotImplementedError):
+            qml.gradients.hadamard_grad(tape, mode=mode)
+
+
+class TestDifferentModes:
+    """Unit test the different modes of gradient."""
+
+    def test_reversed_mode(self):
+        """Test that reversed mode hadamard gradient constructs the correct batch of tapes."""
+
+        tape = qml.tape.QuantumScript(
+            [qml.IsingXY(0.5, wires=(0, 1))], [qml.expval(qml.X(0) + qml.Y(0))], shots=50
+        )
+        batch, fn = qml.gradients.hadamard_grad(tape, mode="reversed")
+        assert len(batch) == 2
+
+        expected_H = qml.IsingXY(0.5, wires=(0, 1)).generator() @ qml.Y(2)
+        expected0 = qml.tape.QuantumScript(
+            [qml.IsingXY(0.5, wires=(0, 1)), qml.H(2), qml.CNOT((2, 0)), qml.H(2)],
+            [qml.expval(expected_H)],
+            shots=50,
+        )
+        qml.assert_equal(batch[0], expected0)
+        expected1 = qml.tape.QuantumScript(
+            [qml.IsingXY(0.5, wires=(0, 1)), qml.H(2), qml.CY((2, 0)), qml.H(2)],
+            [qml.expval(expected_H)],
+            shots=50,
+        )
+        qml.assert_equal(batch[1], expected1)
+
+        assert qml.math.allclose(fn((1.0,)), -2.0)
+
+    def test_direct_mode(self):
+        """Directly test the batch output of applying the direct mode hadamard gradient."""
+        tape = qml.tape.QuantumScript(
+            [qml.IsingXY(0.5, wires=(0, 1))], [qml.expval(qml.X(0) + qml.Y(0))], shots=50
+        )
+        batch, fn = qml.gradients.hadamard_grad(tape, mode="direct")
+        assert len(batch) == 4
+
+        measurement = tape.measurements
+
+        expected0 = qml.tape.QuantumScript(
+            [qml.IsingXY(0.5, wires=(0, 1)), qml.evolve(qml.X(0) @ qml.X(1), np.pi / 4)],
+            measurement,
+            shots=50,
+        )
+        qml.assert_equal(batch[0], expected0)
+        expected1 = qml.tape.QuantumScript(
+            [qml.IsingXY(0.5, wires=(0, 1)), qml.evolve(qml.X(0) @ qml.X(1), -np.pi / 4)],
+            measurement,
+            shots=50,
+        )
+        qml.assert_equal(batch[1], expected1)
+        expected2 = qml.tape.QuantumScript(
+            [qml.IsingXY(0.5, wires=(0, 1)), qml.evolve(qml.Y(0) @ qml.Y(1), np.pi / 4)],
+            measurement,
+            shots=50,
+        )
+        qml.assert_equal(batch[2], expected2)
+        expected3 = qml.tape.QuantumScript(
+            [qml.IsingXY(0.5, wires=(0, 1)), qml.evolve(qml.Y(0) @ qml.Y(1), -np.pi / 4)],
+            measurement,
+            shots=50,
+        )
+        qml.assert_equal(batch[3], expected3)
+
+        assert qml.math.allclose(fn((1.0,)), -0.25)
```

Comment:
```suggestion
        assert qml.math.allclose(fn((1.0,1.0, 1.0, 1.0)), -1)
```
I think `-1` at least 🤷‍♀️ 

Since the batch is of size 4, the result should actually be of size 4.  I may actually add a `strict=True` to the `zip` functions in the postprocessing in the future to catch this kind of thing.

Reply:
yes, I also noticed this, got confused for a second when I saw the postprocessing_fn can take just one value. I think in the reversed test there's also two tapes, I'll update that correspondingly as well

Reply:
done pushing the changes